File: packages/vc/src/multibase.ts
import { base58 } from '@scure/base'

// Multicodec prefix for an Ed25519 public key (0xed), varint-encoded.
const ED25519_PUB = Uint8Array.of(0xed, 0x01)
const ED25519_KEY_LENGTH = 32

/** Multibase base58btc: the letter z followed by the base58 text. */
export function encodeBase58btc(bytes: Uint8Array): string {
  return 'z' + base58.encode(bytes)
}

export function decodeBase58btc(text: string): Uint8Array {
  if (!text.startsWith('z')) throw new Error('not multibase base58btc')
  return base58.decode(text.slice(1))
}

/** An Ed25519 public key as a Multikey publicKeyMultibase value (z6Mk...). */
export function encodeMultikey(publicKey: Uint8Array): string {
  if (publicKey.length !== ED25519_KEY_LENGTH) throw new Error('an Ed25519 public key is 32 bytes')
  const out = new Uint8Array(ED25519_PUB.length + publicKey.length)
  out.set(ED25519_PUB)
  out.set(publicKey, ED25519_PUB.length)
  return encodeBase58btc(out)
}

export function decodeMultikey(text: string): Uint8Array {
  const bytes = decodeBase58btc(text)
  if (bytes[0] !== ED25519_PUB[0] || bytes[1] !== ED25519_PUB[1]) throw new Error('not an Ed25519 Multikey')
  if (bytes.length !== ED25519_PUB.length + ED25519_KEY_LENGTH) throw new Error('an Ed25519 public key is 32 bytes')
  return bytes.slice(ED25519_PUB.length)
}

File: vc-go/multibase/multibase.go
// Package multibase encodes and decodes the two multiformat values Vemphy
// uses: base58btc strings and Ed25519 Multikey public keys.
package multibase

import (
	"crypto/ed25519"
	"errors"
	"strings"

	"github.com/mr-tron/base58"
)

// Multicodec prefix for an Ed25519 public key (0xed), varint-encoded.
var ed25519Pub = []byte{0xed, 0x01}

var (
	ErrNotBase58btc = errors.New("not multibase base58btc")
	ErrNotEd25519   = errors.New("not an Ed25519 Multikey")
	ErrKeyLength    = errors.New("an Ed25519 public key is 32 bytes")
)

// EncodeBase58btc returns the letter z followed by the base58 text.
func EncodeBase58btc(b []byte) string {
	return "z" + base58.Encode(b)
}

func DecodeBase58btc(s string) ([]byte, error) {
	rest, ok := strings.CutPrefix(s, "z")
	if !ok {
		return nil, ErrNotBase58btc
	}
	b, err := base58.Decode(rest)
	if err != nil {
		return nil, ErrNotBase58btc
	}
	return b, nil
}

// EncodeMultikey returns a public key as a publicKeyMultibase value (z6Mk...).
func EncodeMultikey(key ed25519.PublicKey) (string, error) {
	if len(key) != ed25519.PublicKeySize {
		return "", ErrKeyLength
	}
	return EncodeBase58btc(append(append([]byte{}, ed25519Pub...), key...)), nil
}

func DecodeMultikey(s string) (ed25519.PublicKey, error) {
	b, err := DecodeBase58btc(s)
	if err != nil {
		return nil, err
	}
	if len(b) < 2 || b[0] != ed25519Pub[0] || b[1] != ed25519Pub[1] {
		return nil, ErrNotEd25519
	}
	if len(b) != len(ed25519Pub)+ed25519.PublicKeySize {
		return nil, ErrKeyLength
	}
	return ed25519.PublicKey(b[2:]), nil
}
